torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -47,6 +47,35 @@
 FSDP_WRAPPED_MODULE = "_fsdp_wrapped_module"
 
 
+@dataclass
+class ShardingStrategy(Enum):
+    """
+    Specify which sharding strategy will be used for the distributed training.
+    FULL_SHARD: Shard parameters, gradients and optimizer states, this algorithm
+                inserts all_gather before forward and backward computation to gather
+                parameters, also inserts reduce_scatter after backward computation for
+                synchronizing and sharding gradients. Sharded optimizer states are
+                updated locally.
+    SHARD_GRAD_OP: Shard optimizer states and gradients, this algorithm inserts all_gather
```

Comment:
I'm not sure what is the zero-2 implementation in deepspeed. but since we are not free full params after forward, each process could hold the whole model full parameters at the end of forward pass, so it is kind of not sharding parameters here